doc/reference/vm_live_migration_internals.md

```diff
@@ -0,0 +1,96 @@
+# LXD VM live migration
+
+LXD supports {ref}`live-migration-vms` of virtual machines by streaming instance state from a source {abbr}`QEMU (Quick Emulator)` to a target QEMU. VM live migration is supported for all storage pool types.
+
+API extension: `migration_vm_live`
+
+## Conceptual process
+
+The live migration workflow varies depending on the type of storage pool used. The two key scenarios are non-shared storage and shared storage within a cluster (e.g., Ceph). If live state transfer is not supported by a target, a stateful stop is performed prior to migration.
+
+### Live migration for non-shared storage
+
+This process leverages the QEMU built-in Network Block Device (NBD) client-server mechanism to transfer the virtual machine's disk and memory state. Below is an overview of the steps:
+
+1. Set {abbr}`QMP (QEMU Machine Protocol)` capabilities.
+2. Temporary write buffer: Create a temporary {abbr}`QCOW2 (QEMU Copy on Write)` file to store writes that occur in the guest during the initial transfer of the root disk.
+3. Snapshot root disk.
+4. Database record creation: Create the target instance and volume database records.
+5. Storage driver transfer.
+6. Start target QEMU: Launch the QEMU process on the target.
+7. Add NBD node: Attach the NBD target disk as a block node on the source. 
+8. Block device mirror: Perform a block device mirror of the snapshot to the target NBD server.
+9. Pause source VM: Pause the source VM once the block device mirror enters `ready` state.
+10. State transfer: Transfer the memory state over WebSocket.
+11. Enter `pre-switchover` state.
+12. Complete migration: Continue state migration until `completed` state reached.
+
+```{figure} /images/vm_live_migration_flowchart.svg
+Non-Shared Storage Migration
+```
+
+The state transitions during the process are shown below:
+
+```{figure} /images/vm_live_migration_fsa_diagram.svg
+Non-Shared Storage Migration State Transitions
+```
+
+### Intra-cluster member live migration (Ceph shared storage pool)
+
+For shared storage pools such as Ceph, disk data transfer is unnecessary. Instead, the process focuses on transferring the VM state through a dedicated migration socket:
+
+1. Validate cluster state and storage pool readiness.
+2. Transfer the live state data over the migration socket.
+3. Start the target VM using the transferred state.
+
+## Implementation details
+
+### Migration API
+
+Sending a `POST` request to `/1.0/instances/{name}` renames, moves an instance between pools, or migrates an instance to another server. In the push case, the returned operation metadata for migration is a background operation with progress data. For the pull case, it is a WebSocket operation with a number of secrets to be passed to the target server.
+
+### Live migration call stack
+
+Below is a general overview of the key functions of the live migration call stack:
+
+#### [`lxd/lxd/instance_post.go`](https://github.com/canonical/lxd/lxd/instance_post.go)
```

Comment:
This link gives a 404. Seems to be missing the branch name and some other bits needed for file view.

It seems to affect all links to files on GitHub.